src/widgets/scrollbar.rs

```diff
@@ -0,0 +1,208 @@
+use crate::{
+    buffer::Buffer,
+    layout::Rect,
+    style::Style,
+    symbols::{block::FULL, line::DOUBLE_VERTICAL},
+    widgets::Widget,
+};
+
+/// Scrollbar widget for tui-rs library.
+///
+/// This widget can be used to display a scrollbar in a terminal user interface.
+#[derive(Default, Debug, Clone)]
+pub struct Scrollbar<'a> {
+    scroll: f64,
+    style: Style,
+    symbol: &'a str,
+    bar_style: Style,
+    bar_symbol: &'a str,
+}
+
+impl<'a> Scrollbar<'a> {
+    /// Creates a new scrollbar with default styles and symbols.
+    pub fn new(scroll: f64) -> Self {
+        Self {
+            scroll,
+            style: Style::default(),
+            symbol: FULL,
+            bar_style: Style::default(),
+            bar_symbol: DOUBLE_VERTICAL,
+        }
+    }
+
+    /// Sets the scroll position of the scrollbar.
+    pub fn scroll(mut self, scroll: f64) -> Self {
+        if scroll < 0.0 {
+            self.scroll = 0.0;
+        } else {
+            self.scroll = scroll;
+        }
+        self
+    }
+
+    /// Sets the symbol that represents the scroller.
+    pub fn symbol(mut self, symbol: &'a str) -> Self {
+        self.symbol = symbol;
+        self
+    }
+
+    /// Sets the style that is used for the scroller.
+    pub fn style(mut self, style: Style) -> Self {
+        self.style = style;
+        self
+    }
+
+    /// Sets the symbol that represents the bar of the scrollbar.
+    pub fn bar_symbol(mut self, bar_symbol: &'a str) -> Self {
+        self.bar_symbol = bar_symbol;
+        self
+    }
+
+    /// Sets the style that is used for the bar of the scrollbar.
+    pub fn bar_style(mut self, bar_style: Style) -> Self {
+        self.bar_style = bar_style;
+        self
+    }
+}
+
+impl<'a> Widget for Scrollbar<'a> {
+    /// Renders the scrollbar.
+    ///
+    /// The scrollbar is drawn with a fixed width on the right edge of the given area.
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        // If the area is too small, we don't draw the scrollbar.
+        if area.height <= 2 {
+            return;
+        }
+
+        // Calculate the rightmost column for the scrollbar.
+        let right = area.right().saturating_sub(1);
+
+        // If the rightmost column is out of bounds, we don't draw the scrollbar.
+        if right < area.left() {
+            return;
+        };
+
+        // Calculate the top row and the height for the scrollbar.
+        let (top, height) = { (area.top() + 1, area.height.saturating_sub(2)) };
+
+        // Draw the bar of the scrollbar.
+        for y in top..(top + height) {
+            buf.set_string(right, y, self.bar_symbol.clone(), self.bar_style);
```

Comment:
what's the clone() issue that github flags here?

Reply:
I previously had `self.bar_symbol` as a `String` in my personal project but changed it to `&str` before submitting the PR, but I didn't remove the `.clone()`. I think with `&str` the `.clone()` is not required. With the `-D` flag, we get this error:

```
error: using `clone` on a double-reference; this will copy the reference of type `&str` instead of cloning the inner type
Error:   --> src/widgets/scrollbar.rs:91:38
   |
91 |             buf.set_string(right, y, self.bar_symbol.clone(), self.bar_style);
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_double_ref
   = note: `#[deny(clippy::clone_double_ref)]` on by default
help: try dereferencing it
   |
91 |             buf.set_string(right, y, &(*self.bar_symbol).clone(), self.bar_style);
   |                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: or try being explicit if you are sure, that you want to clone a reference
   |
91 |             buf.set_string(right, y, <&str>::clone(self.bar_symbol), self.bar_style);
   |                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```